genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GrpcHeartBeatServiceImpl.java

```diff
@@ -55,20 +57,25 @@
 
     private static final long HEART_BEAT_PERIOD_MILLIS = 5_000L; // TODO make configurable
     private final TaskScheduler taskScheduler;
+    private final MeterRegistry registry;
     private final Map<String, AgentStreamRecord> activeStreamsMap = Maps.newHashMap();
     private final ScheduledFuture<?> sendHeartbeatsFuture;
     private final AgentRoutingService agentRoutingService;
+    private final AtomicInteger connectionsCounter = new AtomicInteger();
 
     GrpcHeartBeatServiceImpl(
         final AgentRoutingService agentRoutingService,
-        @Qualifier("heartBeatServiceTaskScheduler") final TaskScheduler taskScheduler
+        @Qualifier("heartBeatServiceTaskScheduler") final TaskScheduler taskScheduler,
+        final MeterRegistry registry
     ) {
         this.agentRoutingService = agentRoutingService;
         this.taskScheduler = taskScheduler;
+        this.registry = registry;
         this.sendHeartbeatsFuture = this.taskScheduler.scheduleWithFixedDelay(
             this::sendHeartbeats,
             HEART_BEAT_PERIOD_MILLIS
         );
+        this.registry.gauge("genie.grpc.connections.gauge", this.connectionsCounter);
```

Comment:
Instead of all the setting below you should just be able to provide a function to the meter registry with the gauge name and it will sample it for you. In particular it can track collection sizes which is what you're doing here if I'm not mistaken.
[Micrometer Gauge Docs](https://micrometer.io/docs/concepts#_gauges)

Reply:
Fair enough. It's just boilerplate so no big deal. You could use a synchronized map or since it's just a metric and we're not doing something like a financial transaction leave it as a regular map and just take the value as is. If you're planning on using this for OOS or something we'd likely use a rolling count anyway. Either way so long as we're sure we catch all the increment/decrement places this is fine, just thought it was simpler to let the library track the size for us.

Reply:
I'm doing some operations on that map that would not be thread safe with a synchronized map (i.e. lock, inspect, replace, unlock).